### DISCUSSION TITLE
How to filter based on many-to-many entries

### DISCUSSION THREAD

**Discussion by pk992** [pocketbase/pocketbase#2935] [Q&A] [answered]
**Imagine a schema like this:**

![basicM2M](https://github.com/pocketbase/pocketbase/assets/21343763/a0027168-1717-4724-8dae-d08dab93acca)

How can we query all articles with a specific tag? The following seems to be the wrong approach.

```
const records = pb.collection('articles').getFullList({
 filter: `@collection.article_tags(article_id).tag_id.slug = "example-category"`
})
```

Querying the junction table works for a list but results in more problems, because now it's not possible to order articles, for example by most recent (created).
```
const records = pb.collection('article_tags').getFullList({
  filter: `tag_id.slug = "example-category"`,
  expand: 'article_id' 
})
```

I assume it's related to #426 and I was wondering if it's possible currently.

**Comment by ganigeorgiev** [comment, marked as answer]
Using `@collection` in a client-side `filter` expression is not allowed (unless you are authenticated as admin).

Client-side indirect relations `filter` and `sort` at the moment is not supported (_it is in the [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1?pane=issue&itemId=17189183) but remains a low priority_).

The only 2 approaches I can think of at the moment is to either change your DB structure and use a multiple `relation` "tags" field OR send 2 separate requests, something like:

```js
const tags = pb.collection("article_tags").getFullList({
  filter: `tag_id.slug = "example-category"`,
})

const articles = pb.collection("articles").getFullList({
  filter: tags.map((t) => `id="${t.article_id}"`).join("||"),
  sort: "-created",
})
```

**Reply by Zehir** [reply]
Is it fast to use a multiple `relation` ? Btw how is this stored in the database ? it's a comma seperated values ? 
Or instead of a many-to-many table with a view to do queries ?

**Reply by ganigeorgiev** [reply]
@Zehir It will depend on your query and the API rules that you use. Internally multiple values are stored as json array. Accessing and operating with the json is usually fast. DB performance degrades when performing `COUNT` with `JOIN` over large datasets (_but you'll have the same problem if you perform a `JOIN` with a junction collection_). For some example query scenarios you can explore the benchmarks results in https://github.com/pocketbase/benchmarks.

In general, if you plan to have 1k+ relations stored for a single record, a separate junction collection may be better as it will allow more options to query and paginate your data (you'll also have the option to specify plain field indexes for the `relation` fields). The `multiple` options was intended mainly for use cases for linking collections like _tags_, _labels_, _users_, etc. where usually the linked relations are rarely more than 100.

**Reply by pk992** [reply]
@ganigeorgiev Querying the articles based on an array of id's may result in very long fetch URLs, do you see a problem with that approach? If it's ~300 id's.

**Reply by Zehir** [reply]
@ganigeorgiev in my use case is to have an table with around 10k to 100k of row and each row will have a tag list with between 2 and 20 tags from the tag table who there have around 1k of possible tag but most of requests will be find me all row with that specific tag.

Edit : I also have a parent table around 1k entry where each of the 10k row are linked to, see that as post category but a one to many relation. Maybe I could replicate all tag ok the parent and filter thoses first as most of the tags will not change between all post of the same categories

**Reply by ganigeorgiev** [reply]
@pk992 Other than that there could be some limits for the max url string length, performance wise I expect it to be similar to a SQL `IN` statement. Depending on the users demand, eventually we may create a `POST` alias for the list/search endpoint to allow passing the query parameters as body.
